Panic when restarting cube within main repo

<!-- Please describe your issue in English. -->

#### Can be reproduced ?
Yes

#### Steps:
1. Remove `recreateDir` within main of `matrixone`.
2. Restart `matrixone`

#### Expected behavior:


#### Actual behavior:
Panic outputs:

{"level":"info","ts":"2021-08-23T10:12:14.762+0800","caller":"raft/raft.go:1550","msg":"8 switched to configuration voters=(8)"}
2021/08/23 10:12:14.763988 [info] [raftstore]: shard 5 peer 6 ********become leader now********
{"level":"panic","ts":"2021-08-23T10:12:14.763+0800","caller":"raft/log.go:214","msg":"applied(6) is out of range [prevApplied(5), committed(5)]","stacktrace":"go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/pkg/logutil/zap_raft.go:101\ngo.etcd.io/etcd/raft.(*raftLog).appliedTo\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/log.go:214\ngo.etcd.io/etcd/raft.newRaft\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/raft.go:379\ngo.etcd.io/etcd/raft.NewRawNode\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/rawnode.go:48\ngithub.com/matrixorigin/matrixcube/raftstore.newPeerReplica\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/peer_replica.go:208\ngithub.com/matrixorigin/matrixcube/raftstore.createPeerReplica\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/peer_replica.go:159\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).startShards.func1\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:486\ngithub.com/matrixorigin/matrixcube/storage/pebble.(*Storage).Scan\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/storage/pebble/storage.go:181\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).startShards\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:459\ngithub.com/matrixorigin/matrixcube/raftstore.(*store).Start\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:181\ngithub.com/matrixorigin/matrixcube/server.(*Application).Start\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/server/server.go:76\nmatrixone/pkg/vm/engine/aoe/dist.(*driver).Start\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:241\nmain.main\n\t/Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/cmd/db-server/main.go:169\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:255"}
panic: applied(6) is out of range [prevApplied(5), committed(5)]

goroutine 1 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00042c630, {0x0, 0x0, 0x0})
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.uber.org/zap/zapcore/entry.go:230 +0x486
go.uber.org/zap.(*SugaredLogger).log(0xc0000101d8, 0x4, {0x5673f8d, 0x400ed87}, {0xc008411650, 0x5386520, 0xc008528c01}, {0x0, 0x0, 0x0})
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.uber.org/zap/sugar.go:234 +0x13b
go.uber.org/zap.(*SugaredLogger).Panicf(...)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.uber.org/zap/sugar.go:159
go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf(0x0, {0x5673f8d, 0x0}, {0xc008411650, 0x0, 0x0})
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/pkg/logutil/zap_raft.go:101 +0x49
go.etcd.io/etcd/raft.(*raftLog).appliedTo(0xc000438d90, 0x6)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/log.go:214 +0x130
go.etcd.io/etcd/raft.newRaft(0xc00042c6e0)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/raft.go:379 +0x809
go.etcd.io/etcd/raft.NewRawNode(0x8)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/go.etcd.io/etcd/raft/rawnode.go:48 +0x1d
github.com/matrixorigin/matrixcube/raftstore.newPeerReplica(0xc00000d680, 0xc0000f4248, {0x8, 0x1, 0x0, {}, {0x0, 0x0, 0x0}, 0x0})
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/peer_replica.go:208 +0x3fe
github.com/matrixorigin/matrixcube/raftstore.createPeerReplica(0xc00000d680, 0xc0000f4248)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/peer_replica.go:159 +0xb9
github.com/matrixorigin/matrixcube/raftstore.(*store).startShards.func1({0xc0004d6260, 0x0, 0x0}, {0xc00055c1c0, 0x35, 0x35})
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:486 +0x2df
github.com/matrixorigin/matrixcube/storage/pebble.(*Storage).Scan(0xc00051a100, {0x64b08f8, 0x4a27a20, 0xc0001bc4b0}, {0x64b08fa, 0xc008529c20, 0x4cd0985}, 0xc0004710e0, 0x20)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/storage/pebble/storage.go:181 +0x1f6
github.com/matrixorigin/matrixcube/raftstore.(*store).startShards(0xc00000d680)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:459 +0x1a9
github.com/matrixorigin/matrixcube/raftstore.(*store).Start(0xc00000d680)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/raftstore/store.go:181 +0x145
github.com/matrixorigin/matrixcube/server.(*Application).Start(0xc00013d710)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/vendor/github.com/matrixorigin/matrixcube/server/server.go:76 +0x31
matrixone/pkg/vm/engine/aoe/dist.(*driver).Start(0xc000211640)
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:241 +0x25
main.main()
 /Users/yingfeng/codebase/matrixorigin/yingfeng/matrixone/cmd/db-server/main.go:169 +0xbf2


#### Environment:
- Version or commit-id (e.g. v0.1.0 or 8b23a93): 766748e4ee72
- Hardware parameters:
- OS type: Mac
- Others:

--- a/config/config.go
+++ b/config/config.go
@@ -424,6 +424,8 @@ type StoreHeartbeatDataProcessor interface {
 type TestConfig struct {
 	// PeerReplicaSetSnapshotJobWait sleep before set snapshot job
 	PeerReplicaSetSnapshotJobWait time.Duration
+	// SaveDynamicallyShardInitStateWait wait before save dynamically shard init state
+	SaveDynamicallyShardInitStateWait time.Duration
 	// Shards test config for shards
 	Shards map[uint64]*TestShardConfig
 }

--- a/raftstore/keys.go
+++ b/raftstore/keys.go
@@ -117,7 +117,7 @@ func getMetaPrefix(shardID uint64) []byte {
 	return data
 }
 
-func getShardLocaleStateKey(shardID uint64) []byte {
+func getShardLocalStateKey(shardID uint64) []byte {
 	return getMetaKey(shardID, stateSuffix)
 }
 

--- a/raftstore/peer_storage.go
+++ b/raftstore/peer_storage.go
@@ -358,7 +358,7 @@ func compactRaftLog(shardID uint64, state *bhraftpb.RaftApplyState, compactIndex
 }
 
 func loadShardLocalState(shardID uint64, driver storage.MetadataStorage, allowNotFound bool) (*bhraftpb.ShardLocalState, error) {
-	key := getShardLocaleStateKey(shardID)
+	key := getShardLocalStateKey(shardID)
 	v, err := driver.Get(key)
 	if err != nil {
 		logger.Errorf("shard %d load raft state failed with %+v",

--- a/raftstore/store.go
+++ b/raftstore/store.go
@@ -961,14 +961,14 @@ func (s *store) updatePeerState(shard bhmetapb.Shard, state bhraftpb.PeerState, 
 	shardState.Shard = shard
 
 	if wb != nil {
-		return wb.Set(getShardLocaleStateKey(shard.ID), protoc.MustMarshal(shardState))
+		return wb.Set(getShardLocalStateKey(shard.ID), protoc.MustMarshal(shardState))
 	}
 
-	return s.MetadataStorage().Set(getShardLocaleStateKey(shard.ID), protoc.MustMarshal(shardState))
+	return s.MetadataStorage().Set(getShardLocalStateKey(shard.ID), protoc.MustMarshal(shardState))
 }
 
 func (s *store) removePeerState(shard bhmetapb.Shard) error {
-	return s.MetadataStorage().Delete(getShardLocaleStateKey(shard.ID))
+	return s.MetadataStorage().Delete(getShardLocalStateKey(shard.ID))
 }
 
 func (s *store) writeInitialState(shardID uint64, wb *util.WriteBatch) error {

--- a/raftstore/store_bootstrap.go
+++ b/raftstore/store_bootstrap.go
@@ -171,7 +171,7 @@ func (s *store) mustSaveShards(shards ...bhmetapb.Shard) {
 	wb := util.NewWriteBatch()
 	for _, shard := range shards {
 		// shard local state
-		wb.Set(getShardLocaleStateKey(shard.ID), protoc.MustMarshal(&bhraftpb.ShardLocalState{
+		wb.Set(getShardLocalStateKey(shard.ID), protoc.MustMarshal(&bhraftpb.ShardLocalState{
 			Shard: shard,
 		}))
 
@@ -215,7 +215,7 @@ func (s *store) removeInitShards(shards ...bhmetapb.Shard) {
 func (s *store) mustRemoveShards(ids ...uint64) {
 	wb := util.NewWriteBatch()
 	for _, id := range ids {
-		wb.Delete(getShardLocaleStateKey(id))
+		wb.Delete(getShardLocalStateKey(id))
 		wb.Delete(getRaftLocalStateKey(id))
 		wb.Delete(getRaftApplyStateKey(id))
 	}

--- a/raftstore/store_route_handler.go
+++ b/raftstore/store_route_handler.go
@@ -14,6 +14,8 @@
 package raftstore
 
 import (
+	"time"
+
 	"github.com/matrixorigin/matrixcube/pb/bhmetapb"
 )
 
@@ -31,10 +33,15 @@ func (s *store) doDynamicallyCreate(shard bhmetapb.Shard) {
 		return
 	}
 
-	s.addPR(pr)
+	// fix issue 166, we must store the initialization state before pr joins the store
+	// to avoid data problems caused by concurrent modification of this state.
+	if s.cfg.Test.SaveDynamicallyShardInitStateWait > 0 {
+		time.Sleep(s.cfg.Test.SaveDynamicallyShardInitStateWait)
+	}
+	s.mustSaveShards(shard)
 	for _, p := range shard.Peers {
 		s.peers.Store(p.ID, p)
 	}
-	s.mustSaveShards(shard)
+	s.addPR(pr)
 	s.updateShardKeyRange(shard)
 }

--- a/raftstore/store_test.go
+++ b/raftstore/store_test.go
@@ -28,6 +28,7 @@ import (
 	"github.com/matrixorigin/matrixcube/config"
 	"github.com/matrixorigin/matrixcube/pb"
 	"github.com/matrixorigin/matrixcube/pb/bhmetapb"
+	"github.com/matrixorigin/matrixcube/pb/bhraftpb"
 	"github.com/matrixorigin/matrixcube/pb/raftcmdpb"
 	"github.com/matrixorigin/matrixcube/storage"
 	"github.com/matrixorigin/matrixcube/util/leaktest"
@@ -228,6 +229,33 @@ func TestSpeedupAddShard(t *testing.T) {
 
 	id := c.GetShardByIndex(1).ID
 	c.WaitShardStateChangedTo(t, id, metapb.ResourceState_Running, testWaitTimeout)
+}
+
+func TestIssue166(t *testing.T) {
+	c := NewSingleTestClusterStore(t, WithAppendTestClusterAdjustConfigFunc(func(node int, cfg *config.Config) {
+		cfg.Test.SaveDynamicallyShardInitStateWait = time.Second
+		cfg.Customize.CustomInitShardsFactory = func() []bhmetapb.Shard { return []bhmetapb.Shard{{Start: []byte("a"), End: []byte("b")}} }
+	}), WithTestClusterLogLevel("error"))
+	defer c.Stop()
+
+	c.Start()
+	c.WaitLeadersByCount(t, 1, testWaitTimeout)
+
+	_, err := c.stores[0].CreateResourcePool(metapb.ResourcePool{
+		Capacity:    1,
+		Group:       0,
+		RangePrefix: []byte("b"),
+	})
+	assert.NoError(t, err)
+	c.WaitLeadersByCount(t, 2, testWaitTimeout)
+	// make sure slow point
+	time.Sleep(time.Second)
+
+	v, err := c.stores[0].MetadataStorage().Get(getRaftLocalStateKey(c.GetShardByIndex(1).ID))
+	assert.NoError(t, err)
+	state := &bhraftpb.RaftLocalState{}
+	protoc.MustUnmarshal(state, v)
+	assert.Equal(t, uint64(6), state.HardState.Commit)
 }
 
 func createTestWriteReq(id, k, v string) *raftcmdpb.Request {

